src/Evolution/Systems/NewtonianEuler/InternalEnergyDensity.cpp

```diff
@@ -0,0 +1,48 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/Systems/NewtonianEuler/InternalEnergyDensity.hpp"
+
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/ContainerHelpers.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+
+/// \cond
+namespace NewtonianEuler {
+template <typename DataType>
+void internal_energy_density(
+    const gsl::not_null<Scalar<DataType>*> result,
+    const Scalar<DataType>& mass_density,
+    const Scalar<DataType>& specific_internal_energy) noexcept {
+  destructive_resize_components(result, get_size(get(mass_density)));
+  get(*result) = get(mass_density) * get(specific_internal_energy);
+}
+
+template <typename DataType>
+Scalar<DataType> internal_energy_density(
+    const Scalar<DataType>& mass_density,
+    const Scalar<DataType>& specific_internal_energy) noexcept {
+  Scalar<DataType> result{};
+  internal_energy_density(make_not_null(&result), mass_density,
+                          specific_internal_energy);
+  return result;
+}
+
+#define DTYPE(data) BOOST_PP_TUPLE_ELEM(0, data)
+
+#define INSTANTIATE_SCALAR_ARGS(_, data)                             \
+  template void internal_energy_density(                             \
+      const gsl::not_null<Scalar<DTYPE(data)>*> result,              \
+      const Scalar<DTYPE(data)>& mass_density,                       \
+      const Scalar<DTYPE(data)>& specific_internal_energy) noexcept; \
+  template Scalar<DTYPE(data)> internal_energy_density(              \
+      const Scalar<DTYPE(data)>& mass_density,                       \
+      const Scalar<DTYPE(data)>& specific_internal_energy) noexcept;
+
+GENERATE_INSTANTIATIONS(INSTANTIATE_SCALAR_ARGS, (double, DataVector))
```

Comment:
[stylistic nitpick] Now that you've split things and each .cpp file has just one call to GENERATE_INSTANTIATIONS, I'd favor renaming the instantiation macros (through the entire commit) to simply INSTANTIATE instead of the various flavors of INSTANTIATE_BLAH that were necessary before.